pulsar-client/src/main/java/org/apache/pulsar/client/util/MessageIdUtils.java

```diff
@@ -19,11 +19,12 @@
 package org.apache.pulsar.client.util;
 
 import org.apache.pulsar.client.api.MessageId;
+import org.apache.pulsar.client.api.MessageIdAdv;
 import org.apache.pulsar.client.impl.MessageIdImpl;
 
 public class MessageIdUtils {
```

Comment:
This util class was introduced at very early era. It's better to add the util function to `MessageIdAdvUtils` for `MessageIdAdv`.

And I don't think it's worth providing a util function for `MessageIdAdv` in the Pulsar core repo. `MessageIdAdv` is introduced mainly for this reason. The downstream application should cast `MessageId` to `MessageIdAdv` to access the internal fields.  It should not depend on the wrapped APIs from the upstream.

Pulsar does not have the "offset" concept, the same logic is only used in the Kafka sink connector, but it does not call this util method directly. See https://github.com/apache/pulsar/blob/e558cfe9836256065befb3ff6d6043eca10aa5ef/pulsar-io/kafka-connect-adaptor/src/main/java/org/apache/pulsar/io/kafka/connect/KafkaConnectSink.java#L376

To get the offset, it's better to leverage the `AppendIndexMetadataInterceptor` like KoP. See the discussion here: https://github.com/streamnative/kop/issues/290

Reply:
This is also used in third-party connectors, not just limited to the kafka sink connector.
Removing the `offset` from the pulsar-client looks good to me. I think we can mark it with `Deprecate`.

Reply:
Yes. Before removing it, we can mark it with `@Deprecated`.

In 3rd party applications, it's easy to customize its own implementation via `MessageIdAdv`. You can see the example from an early implementation of KoP: https://github.com/streamnative/kop/blob/branch-2.7.4.5/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/utils/MessageIdUtils.java, which is also an example to show the way to compute the offset is not standard.

Reply:
To leave our discussion context here, I created a new PR to deprecate these APIs: https://github.com/apache/pulsar/pull/22747.
There is the same offset implementation in the `FunctionCommon`. I don't deprecate them currently because the connector still needs to use it.